providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -72,435 +40,441 @@ static OSSL_FUNC_keymgmt_import_types_fn mlkem_imexport_types;
         return 0;
+    if (provctx != NULL)
+        libctx = PROV_LIBCTX_OF(provctx);
 
-    key = OPENSSL_zalloc(sizeof(MLKEM768_KEY));
+    key = OPENSSL_zalloc(sizeof(MLKEM_KEY));
     if (key != NULL) {
-        key->keytype = MLKEM_KEY_TYPE_768; /* TODO(ML-KEM) any type */
         key->provctx = provctx;
-        /*
-         * ideally, this is a one-time allocation and ctx that should be within the
-         * provider context: OK to move it there to improve performance?? It would be
-         * the first algorithmspecific context stored: Feels weird (TODO(ML-KEM)).
-         */
-        key->mlkem_ctx = ossl_mlkem_newctx(provctx == NULL ? NULL : PROV_LIBCTX_OF(provctx), NULL);
-        if (key->mlkem_ctx == NULL) {
+        key->vinfo = v;
+        key->ctx = ossl_mlkem_newctx(libctx, NULL);
+        if (key->ctx == NULL) {
             OPENSSL_free(key);
-            key = NULL;
+            return NULL;
         }
     }
-
-    debug_print("MLKEMKM new key = %p\n", key);
     return key;
 }
 
 static void mlkem_free(void *vkey)
 {
-    MLKEM768_KEY *mkey = (MLKEM768_KEY *)vkey;
+    MLKEM_KEY *key = vkey;
+    vinfo_t v = key->vinfo;
 
-    debug_print("MLKEMKM free key %p\n", mkey);
-    if (mkey == NULL)
+    if (key == NULL)
         return;
-    ossl_mlkem_ctx_free(mkey->mlkem_ctx);
-    OPENSSL_free(mkey->encoded_pubkey);
-    OPENSSL_free(mkey->encoded_privkey);
-    OPENSSL_free(mkey);
-}
-
-static int mlkem_has(const void *keydata, int selection)
-{
-    const MLKEM768_KEY *key = keydata;
-    int ok = 0;
 
-    debug_print("MLKEMKM has %p\n", key);
-    if (ossl_prov_is_running() && key != NULL) {
-        /*
-         * ML-KEM keys always have all the parameters they need (i.e. none).
-         * Therefore we always return with 1, if asked about parameters.
-         */
-        ok = 1;
-
-        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0)
-            ok = ok && key->encoded_pubkey != NULL;
-
-        if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0)
-            ok = ok && key->encoded_privkey != NULL;
-    }
-    debug_print("MLKEMKM has result %d\n", ok);
-    return ok;
+    /* Free all key material, zeroing any private parts. */
+    OPENSSL_free(key->pubkey);
+    ossl_mlkem_vcleanse_prvkey(v, &key->prvkey);
+    ossl_mlkem_ctx_free(key->ctx);
+    OPENSSL_free(key);
 }
 
-static int mlkem_match(const void *keydata1, const void *keydata2, int selection)
+static int mlkem_has(const void *vkey, int selection)
 {
-    const MLKEM768_KEY *key1 = keydata1;
-    const MLKEM768_KEY *key2 = keydata2;
-    int ok = 1;
+    static int selectable =
+        OSSL_KEYMGMT_SELECT_PUBLIC_KEY |
+        OSSL_KEYMGMT_SELECT_PRIVATE_KEY;
+    const MLKEM_KEY *key = vkey;
 
-    debug_print("MLKEMKM matching %p and %p\n", key1, key2);
-    if (!ossl_prov_is_running())
+    if (!ossl_prov_is_running() || key == NULL)
         return 0;
+    if ((selection & selectable) == 0)
```

Comment:
why this variable selectable? Why not just use the define's directly?

Reply:
This is cosmetic, `selectable` is a static constant. I found more convenient to place at the top of the function.